add in a client builder that abstracts away connecting to TLS or non-TLS connections and what TLS provider is used.

urkle opened the pull request:
- this allows a more transparent and versatile usage of the library as one can simply compile it as-is and then use the builder to configure where we connect and how we connect without having to be concerned about what type is used for the imap::Client / imap::Session

This is inspired by how the ldap3 library hides-away the connection of ldap, ldaps, or ldapi (unix socket)

IMHO this should just be the main ClientBuilder, however, the methodology used in this PR requires that only either rustls OR nativetls be active, not both. So this PR adds it as a parallel client builder.

Implements issue described in #244 

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/jonhoo/rust-imap/245)
<!-- Reviewable:end -->

tjni commented:
Disclaimer: I'm a rust novice and new to this crate.

I haven't reviewed in detail, but I really like that enabling only one of `native-tls` or `rustls-tls` uses code that does not need to specify which one is used.

Personally, I think this is the overwhelming common case and would vote to make this part of the current builder even if it removes the option of using both TLS features at the same time, unless someone has a need for it.

If we want to support that situation, would it still be possible by requiring a TLS library enum as a parameter to `tls` when multiple TLS library features are enabled?

jonhoo commented on src/conn.rs:
```diff
@@ -0,0 +1,59 @@
+
+/// A generic stream wrapper around either a Tcp or a Tls connection
+#[derive(Debug)]
+pub enum ConnType {
```
Couldn't we just replace this whole thing with
```rust
pub trait ImapConnection: Read + Write {}
impl<T> ImapConnection for T where T: Read + Write {}

pub type ConnType = Box<dyn ImapConnection>;
```
I think that'd be much nicer, and it should also allow the two TLS implementations to co-exist rather than being mutually exclusive.

jonhoo replied:
Furthermore, by doing this, we should be able to make this just be part of the standard builder.

jonhoo commented on src/boxed_client_builder.rs:
```diff
@@ -0,0 +1,216 @@
+        Ok(client)
+    }
+
+    #[cfg(feature = "rustls-tls")]
+    fn build_tls_connection(&self, tcp: TcpStream) -> Result<ConnType> {
+        let mut config = ClientConfig::builder()
+            .with_safe_defaults()
+            .with_root_certificates(CACERTS.clone())
+            .with_no_client_auth();
+        if self.no_tls_verify {
+            let no_cert_verifier = NoCertVerification;
+            config
+                .dangerous()
+                .set_certificate_verifier(Arc::new(no_cert_verifier));
+        }
+        let ssl_conn: RustlsConnector = config.into();
+        Ok(ssl_conn.connect(self.domain.as_ref(), tcp)?.into())
+    }
+
+    #[cfg(feature = "native-tls")]
+    fn build_tls_connection(&self, tcp: TcpStream) -> Result<ConnType> {
+        let mut builder = TlsConnector::builder();
+        if self.no_tls_verify {
+            builder.danger_accept_invalid_certs(true);
+            builder.danger_accept_invalid_hostnames(true);
+        }
+        let ssl_conn = builder.build()?;
+        Ok(TlsConnector::connect(&ssl_conn, self.domain.as_ref(), tcp)?.into())
+    }
```
The way I'd recommend we do this instead is to have something like:
```rust
#[non_exhaustive]
pub enum TlsKind {
  #[cfg(feature = "native-tls")]
  Native
  #[cfg(feature = "rustls-tls")]
  Rust
  #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
  Any
}
```
with
```rust
impl Builder {
  pub fn tls_kind(mut self, kind: TlsKind) -> Self { ... }
  fn build_tls_connection(&self, tcp: TcpStream) -> Result<ConnType> {
    match self.kind {
      #[cfg(feature = "native-tls")]
      TlsKind::Native => self.build_tls_native(),
      #[cfg(feature = "rustls-tls")]
      Rust => self.build_tls_rustls(),
      #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
      Any => self.build_tls_rustls(), // <- we get to choose the default here
    }
  }
}
```

jonhoo reviewed (commented):
Two high-level comments about the design. Overall, I'm definitely a fan of landing something like this!

urkle commented:
@jonhoo I did the first part of the refactoring that uses your Boxed trait suggestion.  Now the code works with both native and rust enabled.

I'll look next at condensing the builders so we have a single builder

urkle commented:
@jonhoo I added another commit that combines the boxed client builder into the now single client builder.

Note a few contract changes happened.

1) the `native_tls()` and `rustls()` methods have been removed as tls_kind replaces that functionality.
2) the old `connect(FnOne)` method has been renamed `connect_tls_with(FnOnce)` I could not find a nice way to get it to behave and convert the `C` to a  `Box<dyn ImapConnection>` the compiler just yelled at me, so it still returns a `Client<C>`
   - if we can find a generic way to allow the user to set custom certs we may be able to just do away with this method completely :)

jonhoo commented on .github/workflows/check.yml:
```diff
@@ -82,7 +82,7 @@ jobs:
         with:
           command: hack
-          args: --feature-powerset check --all-targets
+          args: check --feature-powerset --all-targets
```
Why this change? I though hack arguments had to come before the cargo subcommand?

urkle replied:
All the docs for the cargo hack command show them being called in this order.. "check --feature-powerset etc...).

jonhoo commented on README.md:
```diff
@@ -34,7 +34,7 @@ Below is a basic client example. See the `examples/` directory for more.
 fn fetch_inbox_top() -> imap::error::Result<Option<String>> {
 
-    let client = imap::ClientBuilder::new("imap.example.com", 993).native_tls()?;
+    let client = imap::ClientBuilder::new("imap.example.com", 993).tls().connect()?;
```
I think we should keep the `native_tls` and `rustls` constructors for convenience.

jonhoo replied:
Also, I think those constructors could utilize `connect_tls_with` to avoid the `Box<dyn>` (and thus get a slight performance boost).

urkle replied:
@jonhoo which is why I kinda like the "enum" approach that the Ldap crate used as it removed the need for the box.

And yes, I can add back in the nativeTls / rusttls methods.

jonhoo replied:
When you do, I think we can also avoid changing all these examples

urkle replied:
well, so, I would think we'd want the examples to be in the preferable "non-surprising" syntax. (e.g. it doesn't break if they switch backends)  (which is why I opted for dropping those methods to begin with)

Also, if you are concerned about performance of the Box we can use an enum of 
```rust
enum Connection {
    Tcp(TcpStream),
    NativeTls(NativeTlsStream),
    RustTls(RusTlsStream),
    Custom(Box<ImapConnection>),
}
```

jonhoo replied:
Hmm, maybe. I'm torn, because it's _also_ useful to show folks how they can choose a particular TLS backend to use. I'll go whichever way you prefer here :+1: 

No, `Box<dyn>` is generally going to be better performance than dispatch-through-enum I think.

urkle replied:
We can add examples showing the variations, and those two methods (rustls, and native_tls) that I added back in have doc-comment examples showing usage.

jonhoo commented on src/client_builder.rs:
```diff
@@ -61,35 +129,88 @@ where
     }
 
     /// Use [`STARTTLS`](https://tools.ietf.org/html/rfc2595) for this connection.
-    #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
```
We should still require one of these features to be set, because otherwise `ConnectionMode::StartTls` won't work.

urkle replied:
Which I handled in the connection method by having it return an error staging that no TLS was configured.

Any the reason I removed this #[cfg] was that if I didn't this enumeration would have zero values when no TLS implementation was compiled in.   Leaving this in felt the simplest way to handle that nicely.

the other way would be to completely CFG out this entire enum and tlskind struct member and other functions. (which I can do instead if you prefer that rough)

jonhoo replied:
Why is it a problem for the enum to have no variants?

I'd rather have things fail at compile-time (i.e., make it impossible to write code that tries to `STARTTLS` when there's no TLS available), rather than have it return a runtime error.

urkle replied:
I removed the entire type when no TLS backend is compiled in, and the tls/starttls methods are also removed (based on your later comments).  So the runtime error code is gone now.

jonhoo commented on src/client_builder.rs:
```diff
@@ -1,55 +1,123 @@
+enum ConnectionMode {
+    /// A plain unencrypted Tcp connection
+    Tcp,
+    /// an encrypted TLS connection
+    Tls,
+    /// a start tls connection
+    StartTls,
```
These should both have `#[cfg]`s.

jonhoo commented on src/client_builder.rs:
````diff
@@ -61,35 +129,88 @@ where
+    /// ```
+    pub fn connect(&self) -> Result<Client<ConnType>> {
+        let tcp = TcpStream::connect((self.domain.as_ref(), self.port))?;
+        #[cfg(any(feature = "rustls-tls", feature = "native-tls"))]
````
I think the body of this function would look a lot nicer if structured as:
```rust
match self.mode {
```
with each arm that is only available with TLS carrying a `#[cfg]`.

jonhoo commented on src/client_builder.rs:
```diff
@@ -61,35 +129,88 @@ where
+        let stream: ConnType = if self.mode == ConnectionMode::Tcp {
+            Box::new(tcp)
+        } else {
+            return Err(crate::Error::NoTlsSupport);
```
We should make it impossible for this error to occur.

jonhoo commented on src/client_builder.rs:
```diff
@@ -1,55 +1,123 @@
+    #[cfg(feature = "rustls-tls")]
+    Rust,
+    /// Use whatever backend is available (native over rustls)
+    Any,
```
This should have `#[cfg(any(...))]`.

jonhoo commented on src/client_builder.rs:
````diff
@@ -116,19 +237,23 @@ where
     /// # }
     /// ```
-    pub fn connect<F, C>(&mut self, handshake: F) -> Result<Client<C>>
+    pub fn connect_tls_with<F, C>(&mut self, handshake: F) -> Result<Client<C>>
````
Note that this isn't _necessarily_ for TLS. This can be used for _any_ customization of the underlying I/O channel. I think it should just be called `connect_with`.

jonhoo commented on src/client_builder.rs:
```diff
@@ -116,19 +237,23 @@ where
         C: Read + Write,
     {
-        let tcp = if self.starttls {
+        if self.mode == ConnectionMode::Tcp {
+            panic!("Can only be called with a TLS or StartTLS connection");
+        }
```
This doesn't seem right to me. My expectation would actually be that if someone calls `connect_with` and hasn't called `tls()`, then this should work just fine and just use exactly what comes out of the `FnOnce` as the channel. This suggests, perhaps, that `Tcp` isn't the right name for this mode, and that it should instead be called something like `Direct` or `Plain` or `Unmodified`.

urkle replied:
Per your comment about the function name, I'm just removing this check completely.

jonhoo commented on src/client_builder.rs:
```diff
@@ -141,10 +266,62 @@ where
 
         let mut client = Client::new(tls);
-        if !self.starttls {
+        if self.mode != ConnectionMode::StartTls {
```
I think now that we have `self.mode`, I would expect this to _also_ respect `ConnectionMode::Tls`. That is, if someone did:
```rust
builder.tls().connect_with(|_host, tcp| tcp)
```
I would expect them to end up with a TLS connection. This comes back to the fact that `connect_with` is for wrapping the channel _in any way_, and so does not in and of itself imply that TLS has already been set up. Which also then means that it'd be a mistake to call `.tls().connect_with(|_, tcp| Rustls::new(tcp))`, because you'd end up running TLS-over-TLS.

jonhoo replied:
We should also document this behavior.

urkle replied:
Which that now greatly complicates this method's use.  As if you go with that logic how would StartTLS behave?  Would the user of this method have to initialize starttls? or would the function still perform that behavior?

As your original intent for this method was to "customize" the TLS setup, which one cannot do AFTER the TLS object has been created and wrapped around the raw TCP connection.  Thus this is intended to be done before wrapping.

So, perhaps we should just scrap this method entirely until we come up with a better way to customize the connections?  e.g. a customize_native / customize_rustls ?  that provides the config/builder struct allowing the user to tweak it ?

e.g.
```rust
client_builder.tls().customize_nativetls(|&mut builder| .... modify builder).connect()?;
```

and in our build_native_tls() internal method we do this.

```rust
fn build_tls_native(&self, tcp: TcpStream) -> Result<ConnType> {
        let mut builder = NativeTlsConnector::builder();
        if self.no_tls_verify {
            builder.danger_accept_invalid_certs(true);
            builder.danger_accept_invalid_hostnames(true);
        }
        if Some(customizer) = self.customize_nativetls_callback {
           customizer(builder);
        }
        let ssl_conn = builder.build()?;
        Ok(Box::new(NativeTlsConnector::connect(
            &ssl_conn,
            self.domain.as_ref(),
            tcp,
        )?))
    }
```

We can either have two separate customization hooks OR have a unified with that takes an enum of either the nativetls builder or the rusttls config.   Also if a consumer is customizing the TLS connection do we want to process the defaults always? (loading stock certs?) or add a config option to disable that functionality?

jonhoo replied:
Hmm, I do see your point, but at the same I simultaneously think this method is useful as-is _and_ I think it should do something reasonable in the `client.tls().connect_with` situation. I suppose we could return a runtime error? An alternative would be to do a type-state kind of thing where instead of keeping `mode` as a variable we keep it as a type parameter, but that feels excessive. Maybe just document that it'll return an error if `tls()` is set?

urkle replied:
What error should it return?  ( a new variant?)

jonhoo replied:
Yeah, a new variant seems fine

jonhoo commented on src/conn.rs:
```diff
@@ -0,0 +1,18 @@
+use std::io::{Read, Write};
+
+/// Imap connection trait of a read/write stream
+pub trait ImapConnection: Read + Write + SetReadTimeout {}
```
We should probably [seal](https://rust-lang.github.io/api-guidelines/future-proofing.html#sealed-traits-protect-against-downstream-implementations-c-sealed) this trait.

jonhoo commented on src/error.rs:
```diff
@@ -104,6 +105,8 @@ pub enum Error {
     /// In response to a STATUS command, the server sent OK without actually sending any STATUS
     /// responses first.
     MissingStatusResponse,
+    /// No TLS Backend was compiled in and a TLS connection was attempted
+    NoTlsSupport,
```
I think we should make this impossible.

urkle replied:
it's removed now

jonhoo commented on src/conn.rs:
```diff
@@ -0,0 +1,18 @@
+}
+
+/// A boxed connection type
+pub type ConnType = Box<dyn ImapConnection>;
```
I think a better name for this would be `Connection`.

jonhoo commented on src/client_builder.rs:
```diff
@@ -61,38 +134,91 @@ where
-        })
+    /// Sets the TLS backend to use for this connection.
+    #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+    pub fn tls_kind(&mut self, kind: TlsKind) -> &mut Self {
```
I wonder if this should auto-set `self.mode = Tls` if it's currently set to `Tcp`. At least, it feels weird to have to call `tls_kind` and _also_ `tls`.

urkle replied:
Well, hopefully any "reasonable" IMAP server is properly configured to disallow non-TLS connections in that case :)  ( I know my public facing IMAP servers are configured this way)

So a few things, 
1) if a user did connect to port 993 without calling TLS() it wouldn't connect.
2) The issue with defaulting TLS on is you are assuming they are connecting to port 993?   (thus port 143 wouldn't work).
3) Should we perhaps maybe introspect the port # passed in and default to TLS if 993 is passed? and starttls if port 143 is passed?
4) Or maybe have it default to an "auto mode" where by
   a) if port = 993 then TLS on
   b) otherwise use starttls *IF* the server supports it (e.g. we would introspect the initially advertised capabilities in the greeting, which a server support STARTTLS should be advertising there)

Honestly #4 would provide the nicest use-case as it would be security by default (if possible). and provide the ease of use you are wanting to give in the crate API.  And uses who want to "go beyond" can simply call the explicit `tls()`, `starttls()`, and (new) `raw()` methods to change the connection mode explicitly.


For "connect_with", we'd still have to figure that out.  Maybe changing the contract to something like this.
`connect_with(starttls: bool, handshake: FnOnce)`
Thus we have to explicitly pass whether we are expecting to do starttls or not.  (false if we are doing RAW or full TLS). Thus allowing that the full flexibility, and it would essentially "ignore" the connection mode enum in the connection object.

jonhoo replied:
I like doing number 4!

For `connect_with`, I guess I wonder why we'd want to have a separate `bool` that overrides whatever `starttls` is set to, rather than just respecting what it's set to directly?

urkle replied:
So, I have "auto" mode implemented locally for the main builder.

Now onto connect_with

So, my thinking of why connect_with would take the starttls param is that it's a "manual" connection mode anyway thus it isn't using any of the "knowledge" from the builder other than the port and host.  (e.g. its even ignoring "tls_kind").

if I have it rely on the ConnectionMode instead then the question is what happens in each scenario.

1. Auto
    - if 993 then TLS (leave alone) otherwise connect and "detect" starttls. Though we really need to "tell" the FnOnce what it is receiving in this case (raw or TLS connection)
2. Tcp/Raw
    - just pass the connection through to the callback
3. StartTls
    - upgrade the TLS always with starttls and pass the upgraded connection to the callback
4. Tls
    - just pass the connection through the callback

Thus to make this really work cleanly we'd really need to change the contract of the callback to accept the socket and the mode so that it knows if it has a Tcp/Raw, or TLS-ready connection has been passed.  or do we ONLY call the callback when we have a TLS connnection?

This goes back to what are the use-cases of this method?

Is it to provide TLS customization? (certs, etc..). If so we can provide better methods to handle that (as I explained elsewhere).

Is it to provide full customization of the TCP connection in other ways?  If so, do we really need to provide that in the ClientBuilder and can't the developer just create a TCP connection themselves?

Is the handling of StartTls in this method making it more difficult?

Anyways,

I'm pushing up a commit that

1) adds an "auto"mode as discussed in point 4 above
2) reworks the connect_with to utilize the "Auto" mode.

I'm going to play around a little more later tonight and possible rewrite the `connect` method to use connect_with and make the code smaller.

jonhoo replied:
I think I'm sold, including probably just removing this kind of "provide a handler" method. As you say, for customizing TCP users can just do that before passing the connection to `Client`, and for TLS configuration we can do better anyway.

Part of what sold me here is also that the current release version of `imap` doesn't even have this kind of a method, so it can't be something a lot of people depend on :p

jonhoo commented on src/client_builder.rs:
```diff
@@ -61,38 +134,91 @@ where
+        let stream = match self.mode {
+            ConnectionMode::Tcp => Box::new(tcp),
+            #[cfg(any(feature = "rustls-tls", feature = "native-tls"))]
+            _ => {
+                let tcp = if self.mode == ConnectionMode::StartTls {
```
I think we should just accept a little bit of repetition here and avoid the `_` match followed by an `if` on the same variable. it'll make this easier to find should we ever add a new `mode` in the future, and it'll make the code a little nicer to read I think.

jonhoo commented on src/client_builder.rs:
```diff
@@ -61,38 +134,131 @@ where
-            Ok(TlsConnector::connect(&ssl_conn, domain, tcp)?)
-        })
+    pub fn native_tls(&mut self) -> Result<Client<Connection>> {
+        if self.mode == ConnectionMode::Tcp {
```
It still feels so weird to me that we force `.tls()` here, but _not_ for `.tls_kind()`.

urkle replied:
Not sure what you mean here.

jonhoo replied:
`Client::tls_kind` does not set `self.mode = Tls`. But `Client::native_tls` (which really just changes the kind) _does_ set `self.mode = Tls`. That feels inconsistent.

jonhoo replied:
I agree — the helpers also end up confusing because it's hard to tell that `tls` sets the _mode_, but `native_tls` sets the _kind_. Having just mode and kind is clearer I think.

jonhoo commented on src/client_builder.rs:
```diff
@@ -141,10 +307,62 @@ where
+        }
+    }
+
+    #[cfg(feature = "native-tls")]
+    fn build_tls_any(&self, tcp: TcpStream) -> Result<Connection> {
+        self.build_tls_native(tcp)
+    }
```
How do you feel about making rustls the default instead of native-tls?

urkle replied:
I have no real opinion (other than I HAD to use native_tls for the project I am on as rustls wouldn't connect to 'older' machines)

jonhoo commented on src/client_builder.rs:
```diff
@@ -61,41 +142,125 @@ where
 
-    /// Return a new [`Client`] using a `native-tls` transport.
+    /// Makes a [`Client`] connection using the native TLS backend.
+    /// Forces TLS unless [`ClientBuilder::starttls()`] was called
```
I _think_ rustdoc prefers the first line to have a paragraph break after it. At least iirc, the entire "first paragraph" is considered the "short description", which would include the "Forces TLS" bit as currently written. Same for the method below.

urkle commented:
@jonhoo I rebased, squashed and cleaned things up a bit based on the latest comments.

Can you re-review things?

jonhoo commented on src/client_builder.rs:
```diff
@@ -61,41 +155,132 @@ where
+    /// [`native_tls::TlsConnectorBuilder::danger_accept_invalid_hostnames`],
+    /// [`rustls::ClientConfig::dangerous`]
+    #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+    pub fn danger_no_tls_verify(&mut self, no_tls_verify: bool) -> &mut Self {
+        self.no_tls_verify = no_tls_verify;
```
The double negation here strikes me as unfortunate. Can we name the field and the method just `danger_tls_verify` and `tls_verify`, and have the latter default to `true` instead?

jonhoo replied:
Alternatively, `danger_skip_tls_verify`.

urkle replied:
Skip tls verify sounds better. as the default should be to verify w/o needing to call anything.

jonhoo commented on src/client_builder.rs:
```diff
@@ -1,55 +1,149 @@
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ConnectionMode {
+    /// Automatically detect what connection mode should be used.
+    /// This will use TLS if the port is 993 and StartTLls if available.
```
```suggestion
    /// This will use TLS if the port is 993, and StartTLS if the server says it's available.
```

jonhoo commented on src/client_builder.rs:
```diff
@@ -61,41 +155,132 @@ where
-    /// Return a new [`Client`] using a `native-tls` transport.
+    /// Sets the TLS backend to use for this connection.
+    #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+    pub fn tls_kind(&mut self, kind: TlsKind) -> &mut Self {
```
If we're now getting rid of the helpers, I think we should have this also specifically set `mode` to be `Tls` _if_ it is currently set to `Tcp`.

urkle replied:
I disagree with that one as they could intent mode to be StartTls, thus we really don't know.  And with my proposal of having the default AutoTls (don't allow plain text by default) this becomes even more so unexpected behavior.

jonhoo commented on src/client_builder.rs:
```diff
@@ -114,37 +299,115 @@ where
-        if !self.starttls {
+                    let capabilities = client.capabilities()?;
+                    if capabilities.has(&Capability::Atom("STARTTLS".into())) {
+                        client.run_command_and_check_ok("STARTTLS")?;
+                        let tcp = client.into_inner()?;
+                        Box::new(handshake(self.domain.as_ref(), tcp)?)
```
I think this is wrong if the user hasn't enabled any of the TLS-providing features. It'll call through to `handshake`, but `handshake` then just returns `Ok(tcp)`, even though we've told the server to switch to an encrypted channel. You probably need a `cfg!()` here to not send `STARTTLS` if we don't have TLS capability.

jonhoo commented on src/client_builder.rs:
```diff
@@ -61,41 +155,132 @@ where
+        self.connect_with(|_domain, tcp| {
+            #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+            return self.build_tls_connection(tcp);
+            #[cfg(all(not(feature = "native-tls"), not(feature = "rustls-tls")))]
+            return Ok(tcp);
```
This is a nit, but I'd kind of like for `handshake` to be called something like `tls_handshake`, and then for us to ensure in `connect_with` that we never call it if we don't have any TLS capability features enabled. Which would then make this be a `unreachable!` rather than `return Ok(tcp)`. What do you think?

jonhoo commented on src/client_builder.rs:
```diff
@@ -61,41 +155,132 @@ where
     /// to be used if your TLS setup requires custom work such as adding private CAs
     /// or other specific TLS parameters.
     ///
+    /// Note, if the connection does not end up as TLS, handshake will not be called.
```
This isn't _quite_ true because we may call `handshake` (currently at least) even if no TLS-enabling features are on, which would mean it's called despite the connection _not_ ending up as TLS.

urkle replied:
Which in another comment you suggested getting rid of connect_with entirely.  Do you still want to keep connect_with? (which I agree with the unreachable! comment before and will remove that code-path)

jonhoo commented on src/client_builder.rs:
```diff
@@ -1,55 +1,149 @@
+    /// Use the Rustls backend
+    #[cfg(feature = "rustls-tls")]
+    Rust,
+    /// Use whatever backend is available (rustls used it both are compiled in)
```
```suggestion
    /// Use whatever backend is available (uses rustls if both are available)
```

jonhoo commented on src/client_builder.rs:
```diff
@@ -61,41 +155,132 @@ where
             domain,
             port,
-            starttls: false,
+            mode: ConnectionMode::Auto,
```
I'm tempted to have `mode` default to `ConnectionMode::Tls`, just because these days it really is what most sane clients should be using (even `STARTTLS` has its problems). Users can then always override that behavior if they know they want to connect to a less-secure server. Maybe we have a helper method called `.compatible()` or something that will set `Auto + Any`?

urkle replied:
Most of my clients do starttls by default vs using the SSL only port.   And anyways if they connect to port 993 it'll use Tls full anyway in Auto mode.

Maybe add a new mode that is AutoTls that will connect to any Tls (starttls or tls) by default only (and not non-tls).. and if the user wants to opt-in to anything (even insecure) they can just set it to auto explicitly?

jonhoo commented on src/client_builder.rs:
```diff
@@ -1,55 +1,149 @@
+    /// This will use TLS if the port is 993 and StartTLls if available.
+    Auto,
+    /// A plain unencrypted Tcp connection
+    Tcp,
```
I think I prefer the name `Plaintext` for this. What do you think?

jonhoo reviewed (changes requested):
We're so so close! Thanks for not giving up on this :sweat_smile: 
A few more nits, but they're small.

I also think we might want to add `= Connection` to the `T` generic type parameter on both `Connection` and `Session`, just so that in the (common) case where people use the builder, they don't have to specify `<Connection>` everywhere.

urkle commented:
@jonhoo I rebased + squashed and then added a fixup addressing the latest comments

- added new Error for TlsNotConfigured  (when no tls is compiled in) and StartTlsNotAvailable (server does not provide but was requested)
- added AutoTls mode (default when tls available) which will refuse to connect via Plaintext
  - when tls is NOT available should we default to Plaintext vs Auto?
- renamed Tcp to Plaintext
- more cfg guards for when tls is disabled to return TlsNotConfigured (even in your unreachable suggestion)
- danger_no_tls_verify -> danger_skip_tls_verify
- removed public "helpers" starttls, tls, native_tls, and rusttls

Other items I am wondering about
1) remove connect_with completely you had made one comment about this that if the user want's to set it it up themselves they can simply wrap it in the Client.  Though if we do this, should we expose into_inner() publicly?
2) change mode and tls_kind, and danger_skip_tls_verify (the 3 config methods) not be &mut self but instead be simply mut self (consuming setters).   (`fn mode(mut self, mode: ConnectionMode) -> Self`)
This would make it so we don't need the configuration to be mut in the common case.

bsodmike commented:
Hello! @urkle Running tests locally for your wip branch I am seeing:

```rust
---- auto_tls stdout ----
thread 'auto_tls' panicked at tests/builder_integration.rs:103:31:
called `Result::unwrap()` on an `Err` value: StartTlsNotAvailable
```

Do we expect to enforce [needing StartTLS even in the AutoTLS scenario?](https://github.com/jonhoo/rust-imap/pull/245/files#diff-c0004691670dfac759abb1f0d5b3f13981f62e6065a35c3ff69ff6bc0a5b5a19R101-R103)

urkle commented:
> ```rust
> ---- auto_tls stdout ----
> thread 'auto_tls' panicked at tests/builder_integration.rs:103:31:
> called `Result::unwrap()` on an `Err` value: StartTlsNotAvailable
> ```

@bsodmike the intention here is that if you use AutoTls, then it will ONLY connect via a secured connection.  If you use port 993 then strait Tls, if you use any other port than it'll use StartTls, and if no starttls is supported it'll yield that StartTlsNotAvailable error.

I just pushed up a force commit to correct an issue in the AutoTls branch where if no TLS provider is compiled in it returns the wrong error message, it should have yielded TlsNotConfigured in that scenario.

The intention of the "Auto" mode is that it'll allow connecting to an Imap server via plain text if no starttls is available. Thus that, in general, is not preferred as it is less secure.

bsodmike commented:
> @bsodmike the intention here is that if you use AutoTls, then it will ONLY connect via a secured connection.

So it seems (by default) the `auto_tls` is meant to fail but the test asserts it should pass instead? If that's the case, the test needs patching up :)

jonhoo commented on src/client_builder.rs:
```diff
@@ -61,49 +160,85 @@ where
-            starttls: false,
+            #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+            mode: ConnectionMode::AutoTls,
+            #[cfg(all(not(feature = "native-tls"), not(feature = "rustls-tls")))]
+            mode: ConnectionMode::Auto,
```
One thing I worry about here is that if someone pulls in this dependency with `default-features = false`, they might not ever realize that they're doing plaintext connections. I wonder if the default mode if not TLS feature is enabled should be one that will _always_ error saying "No TLS implementation available". And we should require an _explicit_ `.mode(ConnectionMode::Plaintext)`.

jonhoo replied:
Which, I suppose, is really just saying that I think the default should _always_ be `AutoTls`.

jonhoo commented on src/client_builder.rs:
```diff
@@ -61,49 +160,85 @@ where
     /// or other specific TLS parameters.
     ///
+    /// Note, if the connection does not end up as TLS, handshake will not be called.
+    /// This can happen if connection mode is set to Tcp, or the server does not support starttls.
```
```suggestion
    /// This can happen if connection mode is set to "plaintext", or if the server does not support STARTTLS (and the connection mode is `Auto`).
```

jonhoo reviewed (changes requested):
This is looking really good! Mostly smaller nits now.

As for your questions:

1. I think exposing `.into_inner` publicly is probably a good idea. Now that `connect_with` specifically only takes a thing that initiates a TLS handshake, I don't think it's very useful publicly any more. Let's make it non-`pub`, and then we can open it up if someone specifically asks for it in the future :+1: 
2. I like that idea :+1:

urkle commented on src/client.rs:
```diff
@@ -340,11 +340,24 @@ impl<T: Read + Write> Client<T> {
     ///
     /// This consumes `self` since the Client is not much use without
     /// an underlying transport.
+    #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
```
I think it was because it is only used in that scenario..  So it was to silence warnings:)  but since we are now making this pub I'll remove the cfg.

urkle commented on src/client.rs:
```diff
@@ -340,11 +340,24 @@ impl<T: Read + Write> Client<T> {
+    /// one of the listed capabilities. See [`Capabilities`] for further details.
+    ///
+    /// This allows reading capabilities before authentication.
+    pub fn capabilities(&mut self) -> Result<Capabilities> {
```
Yes, this was so I could query the capabilities BEFORE logging in. (to verify that STARTTLS is available)

jonhoo commented on src/client_builder.rs:
```diff
@@ -1,55 +1,159 @@
+
+/// The connection mode we are going to use
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ConnectionMode {
```
This needs to be marked with `#[non_exhaustive]` since the variants vary (and we may add others in the future).

jonhoo commented on src/client_builder.rs:
```diff
@@ -1,55 +1,159 @@
+    ///
+    /// This will use TLS if the port is 993, and otherwise STARTTLS if available.
+    /// If no TLS communication mechanism is available, the connection will fail.
+    #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
```
I think this needs to not be `cfg`-guarded since we now always set it in `ClientBuilder::new` (in fact the code probably won't compile as-is).

urkle replied:
Should we just make the otherones also not Cfg guarded since we have the proper error code response now when no Tls is compiled in?

jonhoo replied:
No, let's keep those guarded (we can always un-guard them now that it's `#[non_exhaustive]`). It'll still then serve as a signal to folks to remember to enable the TLS features if they specifically try to enforce TLS.

jonhoo reviewed (approved):
Excellent, I think that does it! Just two trivial changes left to make so the code will build, and then I think we're good to go with a rebase!

jonhoo reviewed (approved):
:tada: :tada: :tada: 
Thank you for sticking with it!

jonhoo commented:
Ah, looks like we also now specifically have to set `Auto` (not `AutoTls`) for the greenmail tests, and [fix some doctests](https://github.com/jonhoo/rust-imap/actions/runs/6397840604/job/17366686378?pr=245#step:7:40).

urkle commented:
@jonhoo I rebased + squashed and fixed up the test errors.

for connect_with I simply removed the comments & doc example entirely since it's now an internal method.

urkle commented:
@jonhoo we are green except for the persisting code-coverage one.  I am reviewing those to at least add some coverage to my code changes where possible.

urkle commented:
@jonhoo 100% green!

jonhoo reviewed (approved):
Amazing, thank you! Finally, we can get this in :tada: :heart:

jonhoo commented:
Released in alpha.12 — thanks again for sticking with this for _so_ long!
